### ISSUE TITLE
No data

### ISSUE THREAD

**Issue by noahfraiture** [MacroPower/wakatime_exporter#15]
Hello, I'm not sure it sure it's the right place but I don't succeed to have data on my grafana. In first time I tried to use my own config but I ended up using your compose but I don't have any data on the grafana. Is there something I'm missing ? 
I have a self-host wakapi instance running on docker, I modified the exporter config with `WAKA_SCRAPE_URI: 'http://localhost:3000/api/compat/wakatime/v1'`, added my own api-key and modify the exposed port for grafana to avoid port conflict, but I still don't have any data on my grafana. Do you have any idea why ? I have good data on my selfhost wakapi instance.
Thank you for your consideration

**Comment by MacroPower** [comment]
Hello, could you post your docker-compose.yaml please? I'd be happy to take a look 🙂

**Comment by noahfraiture** [comment]
Ofc, here it is 
```
version: '3'

networks:
  wakatime-exporter:

services:
  wakatime-exporter:
    image: macropower/wakatime-exporter:latest
    ports:
      - '9212:9212'
    environment:
      WAKA_API_KEY: 'my self hosted key'
      WAKA_SCRAPE_URI: 'http://localhost:3000/api/compat/wakatime/v1'
    networks:
      - wakatime-exporter

  grafana:
    image: grafana/grafana:8.1.1
    ports:
      - '3001:3000'
    volumes:

     - ./docs/dashboards:/etc/dashboards
      - ./docs/provisioning/dashboards.yaml:/etc/grafana/provisioning/dashboards/dashboards.yaml
      - ./docs/provisioning/datasources.yaml:/etc/grafana/provisioning/datasources/datasources.yaml
    networks:
      - wakatime-exporter

  prometheus:
    image: prom/prometheus:v2.29.1
    ports:
      - 9090:9090
    volumes:
      - ./docs/example/prometheus.yaml:/etc/prometheus/prometheus.yaml
    command:
      - '--config.file=/etc/prometheus/prometheus.yaml'
    networks:
      - wakatime-exporter```

**Comment by MacroPower** [comment]
Looking at this, unless you swapped out the real URL with `localhost`, `WAKA_SCRAPE_URI: 'http://localhost:3000/api/compat/wakatime/v1'` won't work. In a container, `localhost` references the container/pod's namespace. i.e., localhost for you isn't the same as localhost for the container. This might be helpful https://stackoverflow.com/questions/24319662/from-inside-of-a-docker-container-how-do-i-connect-to-the-localhost-of-the-mach

If that doesn't seem right or you just swapped out the URL before posting:

- Does wakatime-exporter give you any interesting logs? e.g. some "collector failed" messages.
- Does the Prometheus instance succeed in scraping the wakatime-exporter target?
- If so, are there metrics in Prometheus?

**Comment by noahfraiture** [comment]
Right, I changed that by including my wakapi instance in the compose on the same network but it still doens't work :(
I indeed see error in the collection of data of exporter in logs 
```
level=info ts=2023-12-27T15:15:10.115Z caller=main.go:100 msg="Starting wakatime_exporter" version="(version=0.1.0, branch=master, revision=345c1d881da034127ec48eaa284e6d72d72119d0)"
level=info ts=2023-12-27T15:15:10.115Z caller=main.go:101 msg="Build context" build_context="(go=go1.14.6, user=root@55c72154720f, date=20201003-23:59:46)"
level=info ts=2023-12-27T15:15:10.115Z caller=utils.go:101 msg="Enabled collectors"
level=info ts=2023-12-27T15:15:10.115Z caller=utils.go:108 collector=all-time
level=info ts=2023-12-27T15:15:10.115Z caller=utils.go:108 collector=goal
level=info ts=2023-12-27T15:15:10.115Z caller=utils.go:108 collector=leader
level=info ts=2023-12-27T15:15:10.115Z caller=utils.go:108 collector=summary
level=info ts=2023-12-27T15:15:10.116Z caller=main.go:126 msg="Listening on" address=:9212
level=info ts=2023-12-27T15:15:22.759Z caller=common.go:58 collector=leader msg="Scraping Wakatime" path=leaders url="http://wakapi:3000/api/compat/wakatime/v1/leaders?cache=false"
level=info ts=2023-12-27T15:15:22.759Z caller=common.go:58 collector=summary msg="Scraping Wakatime" path=summaries url="http://wakapi:3000/api/compat/wakatime/v1/users/current/summaries?cache=false&end=today&start=today"
level=info ts=2023-12-27T15:15:22.760Z caller=common.go:58 collector=goal msg="Scraping Wakatime" path=goals url="http://wakapi:3000/api/compat/wakatime/v1/users/current/goals?cache=false"
level=info ts=2023-12-27T15:15:22.760Z caller=common.go:58 collector=all-time msg="Scraping Wakatime" path=all_time_since_today url="http://wakapi:3000/api/compat/wakatime/v1/users/current/all_time_since_today?cache=false"
level=error ts=2023-12-27T15:15:22.763Z caller=collector.go:175 msg="collector failed" name=goal duration_seconds=0.003261524 err="HTTP status 404"
level=error ts=2023-12-27T15:15:22.764Z caller=collector.go:175 msg="collector failed" name=summary duration_seconds=0.005120615 err="HTTP status 401"
level=error ts=2023-12-27T15:15:22.765Z caller=collector.go:175 msg="collector failed" name=all-time duration_seconds=0.005342402 err="HTTP status 401"
level=info ts=2023-12-27T15:15:22.766Z caller=leader.go:74 collector=leader msg="Collecting rank from Wakatime" page=1 updated=0001-01-01T00:00:00Z
level=info ts=2023-12-27T15:16:22.759Z caller=common.go:58 collector=summary msg="Scraping Wakatime" path=summaries url="http://wakapi:3000/api/compat/wakatime/v1/users/current/summaries?cache=false&end=today&start=today"
level=info ts=2023-12-27T15:16:22.760Z caller=common.go:58 collector=all-time msg="Scraping Wakatime" path=all_time_since_today url="http://wakapi:3000/api/compat/wakatime/v1/users/current/all_time_since_today?cache=false"
level=info ts=2023-12-27T15:16:22.760Z caller=common.go:58 collector=leader msg="Scraping Wakatime" path=leaders url="http://wakapi:3000/api/compat/wakatime/v1/leaders?cache=false"
level=info ts=2023-12-27T15:16:22.760Z caller=common.go:58 collector=goal msg="Scraping Wakatime" path=goals url="http://wakapi:3000/api/compat/wakatime/v1/users/current/goals?cache=false"
level=error ts=2023-12-27T15:16:22.762Z caller=collector.go:175 msg="collector failed" name=goal duration_seconds=0.002299195 err="HTTP status 404"
level=error ts=2023-12-27T15:16:22.762Z caller=collector.go:175 msg="collector failed" name=summary duration_seconds=0.003038383 err="HTTP status 401"
level=error ts=2023-12-27T15:16:22.764Z caller=collector.go:175 msg="collector failed" name=all-time duration_seconds=0.003998552 err="HTTP status 401"
level=info ts=2023-12-27T15:16:22.764Z caller=leader.go:74 collector=leader msg="Collecting rank from Wakatime" page=1 updated=0001-01-01T00:00:00Z
```

And my configuration file
```
version: '3'

networks:
  wakatime-exporter:

volumes:
  wakapi-data:

services:
  wakatime-exporter:
    image: macropower/wakatime-exporter:latest
    ports:
      - '9212:9212'
    environment:
      WAKA_API_KEY: 'mykey'
      WAKA_SCRAPE_URI: 'http://wakapi:3001/api/compat/wakatime/v1'
    depends_on:
      - wakapi
    networks:
      - wakatime-exporter
    restart: always

  grafana:
    image: grafana/grafana:8.1.1
    ports:
      - '3000:3000'
    volumes:
      - ./docs/dashboards:/etc/dashboards
      - ./docs/provisioning/dashboards.yaml:/etc/grafana/provisioning/dashboards/dashboards.yaml
      - ./docs/provisioning/datasources.yaml:/etc/grafana/provisioning/datasources/datasources.yaml
    networks:
      - wakatime-exporter
    restart: always

  prometheus:
    image: prom/prometheus:v2.29.1
    ports:
      - 9090:9090
    volumes:
      - ./docs/example/prometheus.yaml:/etc/prometheus/prometheus.yaml
    command:
      - '--config.file=/etc/prometheus/prometheus.yaml'
    networks:
      - wakatime-exporter
    restart: always

  wakapi:
    image: ghcr.io/muety/wakapi:latest
    container_name: wakapi
    ports:
      - 3001:3001
    environment:
      WAKAPI_PASSWORD_SALT: 'f44felyAZAvWE7OwKT7LwEOVTVx0r84X' # RANDOMLY GENERATED
      WAKAPI_PORT: 3001
    volumes:
      - wakapi-data:/data
    networks:
      - wakatime-exporter
    restart: always
```
Edit: I changed the port to avoid conflit but same problem

**Comment by MacroPower** [comment]
It seems like you're getting both 404 and 401 errors. The 404 errors are more or less expected with Wakapi, see #14, basically 1 of the endpoints isn't implemented, no big deal. The 401 errors on the other hand, are an issue. That makes me think that the `WAKA_API_KEY` you're passing isn't correct.

**Comment by noahfraiture** [comment]
Ok yeah of course. The problem is that running this compose recreate the container and thus create a new key. And I can't put a key in the compose that has not been already created. I see two possibility here 

- Modify wakapi to hard code a key, since I'm in localhost it's not a problem but when I'll want to put it on a home server it can be a security issue
- Modify in the exporter container the key once it has been created. I'll search if it's possible to enter the exporter with bash or other shell to export there the value

Thank you for your help

Edit : I went in the docker with sh and replace the env var but it wasn't persistent accross session so I modified the docker compose with the correct and rerun it. It only recreated the exporter so my api key didn't change, so it basically works I think, there's too few data to see every graph I guess but I see once at least, so it must work or I'll check grafana and so.
Now I'll to import data to have old data in grafana if possible
